verifiable-db/src/query/merkle_path.rs

```diff
@@ -291,37 +349,344 @@ where
+            &empty_hash,
+        );
+        let end_node = NodeInfoTarget {
+            embedded_tree_hash: end_node_tree_hash,
+            child_hashes: [left_child_hash, right_child_hash],
+            value: end_node_value,
+            min: end_node_info.node_min.clone(),
+            max: end_node_info.node_max.clone(),
+        };
+        let end_node_hash = end_node.compute_node_hash(b, index_id);
+        if let MerklePathWires::MerklePathWithNeighbors(out_targets) =
+            build_common(b, end_node_hash, index_id, true, Some(end_node_info))
+        {
+            out_targets
+        } else {
+            unreachable!()
+        }
+    }
+
+    pub fn assign(
+        &self,
+        pw: &mut PartialWitness<F>,
+        wires: &MerklePathWithNeighborsTargetInputs<MAX_DEPTH>,
+    ) {
+        self.path_gadget.assign(pw, &wires.path_inputs);
+        pw.set_u256_target_arr(
+            &[
+                wires.end_node_inputs.node_min.clone(),
+                wires.end_node_inputs.node_max.clone(),
+            ],
+            &[self.end_node_min, self.end_node_max],
+        );
+        pw.set_bool_target(
+            wires.end_node_inputs.left_child_exists,
+            self.end_node_children[0].is_some(),
+        );
+        pw.set_bool_target(
+            wires.end_node_inputs.right_child_exists,
+            self.end_node_children[1].is_some(),
+        );
+        let left_child_info = self.end_node_children[0].unwrap_or_default();
+        let right_child_info = self.end_node_children[1].unwrap_or_default();
+        wires
+            .end_node_inputs
+            .left_child_info
+            .set_target(pw, &left_child_info);
+        wires
+            .end_node_inputs
+            .right_child_info
+            .set_target(pw, &right_child_info);
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Enum employed as output type for `build_common` method, which allows to implement the core logic
+/// of `MerklePathGadget` and `MerklePathGadgetWithNeighbors` in a single method, hence avoiding
+/// code duplication
+enum MerklePathWires<const MAX_DEPTH: usize>
+where
+    [(); MAX_DEPTH - 1]:,
+{
+    MerklePath(MerklePathTarget<MAX_DEPTH>),
+    MerklePathWithNeighbors(MerklePathWithNeighborsTarget<MAX_DEPTH>),
+}
+
+/// Build `MerklePathWires`, building `MerklePathWithNeighborsTarget` if the input flag `with_neighbors`
+/// is true, building `MerkelPathTarget` otherwise. The required inputs are:
+/// - `end_node`: The hash of the first node in the path
+/// - `index_id`: Integer identifier of the index column to be placed in the hash
+///     of the nodes of the path
+/// - `end_node_info` : Data about `end_node`; must be provided only if `with_neighbors == true`
+fn build_common<const MAX_DEPTH: usize>(
+    b: &mut CircuitBuilder<F, D>,
+    end_node: HashOutTarget,
+    index_id: Target,
+    with_neighbors: bool,
+    end_node_info: Option<EndNodeInputs>,
+) -> MerklePathWires<MAX_DEPTH>
```

Comment:
I initially thought also about what you suggested but I was unsure whether the `build_path` method would have been well-defined or not. But probably it is since you suggested this structure too :) Done in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/388/commits/600a6e99a6c28e109943dfc0df2805a9c826f1f8.